crates/stackable-operator/src/commons/listener.rs

```diff
@@ -197,14 +204,47 @@ pub struct ListenerIngress {
     pub ports: BTreeMap<String, i32>,
 }
 
+/// The type of a given address.
 #[derive(Serialize, Deserialize, Clone, Copy, Debug, JsonSchema, PartialEq, Eq)]
 #[serde(rename_all = "PascalCase")]
 pub enum AddressType {
+    /// A resolvable DNS hostname.
     Hostname,
+
+    /// A resolved IP address.
     #[serde(rename = "IP")]
     Ip,
 }
 
+/// A mode for deciding the preferred [`AddressType`].
+///
+/// These can vary depending on the rest of the [`ListenerClass`].
+#[derive(Serialize, Deserialize, Clone, Copy, Debug, JsonSchema, PartialEq, Eq)]
+pub enum PreferredAddressType {
+    /// Like [`AddressType::Hostname`], but prefers [`AddressType::Ip`] for [`ServiceType::NodePort`], since their hostnames are less likely to be resolvable.
+    HostnameConservative,
+
+    // Like the respective variants of AddressType. Ideally we would refer to them instead of copy/pasting, but that breaks due to upstream issues:
+    // - https://github.com/GREsau/schemars/issues/222
+    // - https://github.com/kube-rs/kube/issues/1622
+    Hostname,
+    #[serde(rename = "IP")]
+    Ip,
+}
+
+impl PreferredAddressType {
+    pub fn resolve(self, listener_class: &ListenerClassSpec) -> AddressType {
+        match self {
+            PreferredAddressType::HostnameConservative => match listener_class.service_type {
+                ServiceType::NodePort => AddressType::Ip,
+                _ => AddressType::Hostname,
+            },
+            PreferredAddressType::Hostname => AddressType::Hostname,
+            PreferredAddressType::Ip => AddressType::Ip,
+        }
```

Comment:
We could also match on the tuple, but no strong opinion
```suggestion
        match (self, listener_class.service_type) {
            (PreferredAddressType::Hostname, _) => AddressType::Hostname,
            (PreferredAddressType::Ip, _) => AddressType::Ip,
            (PreferredAddressType::HostnameConservative, ServiceType::NodePort) => AddressType::Ip,
            (PreferredAddressType::HostnameConservative, _) => AddressType::Hostname,
        }
```

Reply:
I originally considered this, but decided to split it in order to emphasize that any new pseudotypes might depend on other properties instead.

Reply:
Makes sense